[CI] AsyncBulkByScrollActionTests testScrollResponseBatchingBehavior failing

**Build scan:**
https://gradle-enterprise.elastic.co/s/h7gdqkimbf4tg/tests/:modules:reindex:test/org.elasticsearch.reindex.AsyncBulkByScrollActionTests/testScrollResponseBatchingBehavior


**Reproduction line:**
```
./gradlew ':modules:reindex:test' --tests "org.elasticsearch.reindex.AsyncBulkByScrollActionTests.testScrollResponseBatchingBehavior" -Dtests.seed=5638EFEF212A5042 -Dtests.locale=nl-BE -Dtests.timezone=America/Atka -Druntime.java=22
```

**Applicable branches:**
main

**Reproduces locally?:**
Didn't try

**Failure history:**
[Failure dashboard for `org.elasticsearch.reindex.AsyncBulkByScrollActionTests#testScrollResponseBatchingBehavior`](https://es-delivery-stats.elastic.dev/app/dashboards#/view/dcec9e60-72ac-11ee-8f39-55975ded9e63?_g=(refreshInterval:(pause:!t,value:60000),time:(from:now-7d%2Fd,to:now))&_a=(controlGroupInput:(chainingSystem:HIERARCHICAL,controlStyle:twoLine,ignoreParentSettings:(ignoreFilters:!f,ignoreQuery:!f,ignoreTimerange:!f,ignoreValidations:!t),panels:('0c0c9cb8-ccd2-45c6-9b13-96bac4abc542':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:task.keyword,grow:!t,id:'0c0c9cb8-ccd2-45c6-9b13-96bac4abc542',searchTechnique:wildcard,selectedOptions:!(),singleSelect:!t,title:'Gradle%20Task',width:medium),grow:!t,order:0,type:optionsListControl,width:small),'144933da-5c1b-4257-a969-7f43455a7901':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:name.keyword,grow:!t,id:'144933da-5c1b-4257-a969-7f43455a7901',searchTechnique:wildcard,selectedOptions:!('testScrollResponseBatchingBehavior'),title:Test,width:medium),grow:!t,order:2,type:optionsListControl,width:medium),'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850':(explicitInput:(dataViewId:fbbdc689-be23-4b3d-8057-aa402e9ed0c5,enhancements:(),fieldName:className.keyword,grow:!t,id:'4e6ad9d6-6fdc-4fcc-bf1a-aa6ca79e0850',searchTechnique:wildcard,selectedOptions:!('org.elasticsearch.reindex.AsyncBulkByScrollActionTests'),title:Suite,width:medium),grow:!t,order:1,type:optionsListControl,width:medium)))))


**Failure excerpt:**
```
java.lang.AssertionError: cannot complete future on thread Thread[#70,elasticsearch[testScrollResponseBatchingBehavior][generic][T#7],5,TGRP-AsyncBulkByScrollActionTests] with waiter on thread Thread[#69,elasticsearch[testScrollResponseBatchingBehavior][generic][T#6],5,TGRP-AsyncBulkByScrollActionTests], could deadlock if pool was full
	at java.base/java.util.concurrent.locks.LockSupport.parkNanos(LockSupport.java:410)
	at java.base/java.util.concurrent.LinkedTransferQueue$DualNode.await(LinkedTransferQueue.java:452)
	at java.base/java.util.concurrent.LinkedTransferQueue.xfer(LinkedTransferQueue.java:613)
	at java.base/java.util.concurrent.LinkedTransferQueue.poll(LinkedTransferQueue.java:1268)
	at java.base/java.util.concurrent.ThreadPoolExecutor.getTask(ThreadPoolExecutor.java:1069)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1130)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
	at java.base/java.lang.Thread.run(Thread.java:1570)

  at __randomizedtesting.SeedInfo.seed([5638EFEF212A5042]:0)
  at org.elasticsearch.action.support.PlainActionFuture.assertCompleteAllowed(PlainActionFuture.java:416)
  at org.elasticsearch.action.support.PlainActionFuture.onFailure(PlainActionFuture.java:42)
  at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.lambda$finishHim$3(AbstractAsyncBulkByScrollAction.java:598)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:917)
  at org.elasticsearch.index.reindex.ClientScrollableHitSource.cleanup(ClientScrollableHitSource.java:127)
  at org.elasticsearch.index.reindex.ScrollableHitSource.lambda$close$2(ScrollableHitSource.java:115)
  at org.elasticsearch.index.reindex.ClientScrollableHitSource$2.onResponse(ClientScrollableHitSource.java:114)
  at org.elasticsearch.index.reindex.ClientScrollableHitSource$2.onResponse(ClientScrollableHitSource.java:110)
  at org.elasticsearch.reindex.AsyncBulkByScrollActionTests$MyMockClient.doExecute(AsyncBulkByScrollActionTests.java:1092)
  at org.elasticsearch.client.internal.support.AbstractClient.execute(AbstractClient.java:138)
  at org.elasticsearch.client.internal.support.AbstractClient.clearScroll(AbstractClient.java:363)
  at org.elasticsearch.index.reindex.ClientScrollableHitSource.clearScroll(ClientScrollableHitSource.java:110)
  at org.elasticsearch.index.reindex.ScrollableHitSource.close(ScrollableHitSource.java:115)
  at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.finishHim(AbstractAsyncBulkByScrollAction.java:588)
  at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction.finishHim(AbstractAsyncBulkByScrollAction.java:576)
  at org.elasticsearch.reindex.AbstractAsyncBulkByScrollAction$1.onFailure(AbstractAsyncBulkByScrollAction.java:373)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.onFailure(ThreadContext.java:967)
  at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:28)
  at org.elasticsearch.common.util.concurrent.RunOnce.run(RunOnce.java:40)
  at org.elasticsearch.index.reindex.WorkerBulkByScrollTaskState$DelayedPrepareBulkRequest.lambda$new$0(WorkerBulkByScrollTaskState.java:254)
  at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingRunnable.run(ThreadContext.java:917)
  at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1144)
  at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:642)
  at java.lang.Thread.run(Thread.java:1570)

```

--- a/docs/changelog/110361.yaml
+++ b/docs/changelog/110361.yaml
@@ -1,0 +1,7 @@
+pr: 110361
+summary: Don't detect `PlainActionFuture` deadlock on concurrent complete
+area: Distributed
+type: bug
+issues:
+ - 110181
+ - 110360

--- a/server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java
+++ b/server/src/main/java/org/elasticsearch/action/support/PlainActionFuture.java
@@ -379,7 +379,11 @@ public class PlainActionFuture<T> implements ActionFuture<T>, ActionListener<T> 
             } else if (getState() == COMPLETING) {
                 // If some other thread is currently completing the future, block until
                 // they are done so we can guarantee completion.
-                acquireShared(-1);
+                // Don't use acquire here, to prevent false-positive deadlock detection
+                // when multiple threads from the same pool are completing the future
+                while (isDone() == false) {
+                    Thread.onSpinWait();
+                }
             }
             return doCompletion;
         }

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -14,13 +14,21 @@ import org.elasticsearch.common.util.concurrent.UncategorizedExecutionException;
 import org.elasticsearch.core.Assertions;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.test.ESTestCase;
+import org.elasticsearch.threadpool.TestThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;
 
+import java.util.ArrayList;
+import java.util.List;
 import java.util.concurrent.CancellationException;
+import java.util.concurrent.CompletionService;
+import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
+import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
 
 public class PlainActionFutureTests extends ESTestCase {
 
@@ -140,6 +148,66 @@ public class PlainActionFutureTests extends ESTestCase {
         assertCapturesInterrupt(() -> future.get(10, TimeUnit.SECONDS));
         assertPropagatesInterrupt(future::actionGet);
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
+    }
+
+    public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
+        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        try (TestThreadPool tp = new TestThreadPool(getTestName())) {
+            final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+            final int threads = 4;
+            final AtomicBoolean running = new AtomicBoolean(true);
+            final CountDownLatch startBarrier = new CountDownLatch(threads + 1);
+            final CompletionService<?> cs = new ExecutorCompletionService<>(tp.generic());
+            final List<Future<?>> futures = new ArrayList<>();
+            for (int i = 0; i < threads; i++) {
+                futures.add(cs.submit(new FutureCompleter(i, startBarrier, futureReference, running), null));
+            }
+            startBarrier.countDown();
+            safeAwait(startBarrier);
+            cs.poll(500, TimeUnit.MILLISECONDS);
+            running.set(false);
+            futures.forEach(ESTestCase::safeGet);
+        }
+    }
+
+    private static class FutureCompleter implements Runnable {
+        private final int number;
+        private final AtomicReference<PlainActionFuture<?>> future;
+        private final AtomicBoolean running;
+        private final CountDownLatch startBarrier;
+        private volatile boolean failed = false;
+
+        private FutureCompleter(
+            int number,
+            CountDownLatch startBarrier,
+            AtomicReference<PlainActionFuture<?>> future,
+            AtomicBoolean running
+        ) {
+            this.number = number;
+            this.future = future;
+            this.running = running;
+            this.startBarrier = startBarrier;
+        }
+
+        @Override
+        public void run() {
+            try {
+                startBarrier.countDown();
+                safeAwait(startBarrier);
+                while (running.get()) {
+                    if (number == 0) {
+                        future.set(new PlainActionFuture<>());
+                    }
+                    future.get().onResponse(null);
+                }
+            } finally {
+                if (running.get()) {
+                    // We failed
+                    failed = true;
+                }
+                running.set(false);
+            }
+        }
     }
 
     private static void assertCancellation(ThrowingRunnable runnable) {

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -15,16 +15,16 @@ import org.elasticsearch.core.Assertions;
 import org.elasticsearch.core.TimeValue;
 import org.elasticsearch.test.ESTestCase;
 import org.elasticsearch.threadpool.TestThreadPool;
+import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;
 
 import java.util.ArrayList;
 import java.util.List;
 import java.util.concurrent.CancellationException;
-import java.util.concurrent.CompletionService;
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorCompletionService;
+import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -150,59 +150,37 @@ public class PlainActionFutureTests extends ESTestCase {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
-    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() throws InterruptedException {
+    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
         assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
         try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
             final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
-            final int threads = 4;
+            final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+            final var threadCount = threadPool.info(executorName).getMax();
             final AtomicBoolean running = new AtomicBoolean(true);
-            final CountDownLatch startLatch = new CountDownLatch(threads + 1);
-            final CompletionService<?> completionService = new ExecutorCompletionService<>(threadPool.generic());
+            final CountDownLatch startLatch = new CountDownLatch(threadCount + 1);
             final List<Future<?>> futures = new ArrayList<>();
-            for (int i = 0; i < threads; i++) {
-                futures.add(completionService.submit(new FutureCompleter(i, startLatch, futureReference, running), null));
+            final ExecutorService executor = threadPool.executor(executorName);
+            for (int i = 1; i < threadCount; i++) {
+                futures.add(executor.submit(() -> {
+                    startLatch.countDown();
+                    safeAwait(startLatch);
+                    while (running.get()) {
+                        futureReference.get().onResponse(null);
+                    }
+                }, null));
             }
+            executor.submit(() -> {
+                startLatch.countDown();
+                safeAwait(startLatch);
+                while (running.get()) {
+                    futureReference.set(new PlainActionFuture<>());
+                }
+            }, null);
             startLatch.countDown();
             safeAwait(startLatch);
-            completionService.poll(20, TimeUnit.MILLISECONDS);
+            safeSleep(20);
             running.set(false);
             futures.forEach(ESTestCase::safeGet);
-        }
-    }
-
-    private static class FutureCompleter implements Runnable {
-        private final int number;
-        private final AtomicReference<PlainActionFuture<?>> future;
-        private final AtomicBoolean running;
-        private final CountDownLatch startBarrier;
-
-        private FutureCompleter(
-            int number,
-            CountDownLatch startBarrier,
-            AtomicReference<PlainActionFuture<?>> future,
-            AtomicBoolean running
-        ) {
-            this.number = number;
-            this.future = future;
-            this.running = running;
-            this.startBarrier = startBarrier;
-        }
-
-        @Override
-        public void run() {
-            try {
-                startBarrier.countDown();
-                safeAwait(startBarrier);
-                while (running.get()) {
-                    if (number == 0) {
-                        future.set(new PlainActionFuture<>());
-                    }
-                    future.get().onResponse(null);
-                }
-            } finally {
-                // In the event we failed, stop all the other FutureCompleter instances
-                running.set(false);
-            }
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -150,7 +150,7 @@ public class PlainActionFutureTests extends ESTestCase {
         assertPropagatesInterrupt(() -> future.actionGet(10, TimeUnit.SECONDS));
     }
 
-    public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
+    public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() throws InterruptedException {
         assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
         try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
             final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -164,7 +164,7 @@ public class PlainActionFutureTests extends ESTestCase {
             }
             startBarrier.countDown();
             safeAwait(startBarrier);
-            cs.poll(500, TimeUnit.MILLISECONDS);
+            cs.poll(250, TimeUnit.MILLISECONDS);
             running.set(false);
             futures.forEach(ESTestCase::safeGet);
         }
@@ -175,7 +175,6 @@ public class PlainActionFutureTests extends ESTestCase {
         private final AtomicReference<PlainActionFuture<?>> future;
         private final AtomicBoolean running;
         private final CountDownLatch startBarrier;
-        private volatile boolean failed = false;
 
         private FutureCompleter(
             int number,
@@ -201,10 +200,7 @@ public class PlainActionFutureTests extends ESTestCase {
                     future.get().onResponse(null);
                 }
             } finally {
-                if (running.get()) {
-                    // We failed
-                    failed = true;
-                }
+                // In the event we failed, stop all the other FutureCompleter instances
                 running.set(false);
             }
         }

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -164,7 +164,7 @@ public class PlainActionFutureTests extends ESTestCase {
             }
             startBarrier.countDown();
             safeAwait(startBarrier);
-            cs.poll(250, TimeUnit.MILLISECONDS);
+            cs.poll(20, TimeUnit.MILLISECONDS);
             running.set(false);
             futures.forEach(ESTestCase::safeGet);
         }

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -149,21 +149,27 @@ public class PlainActionFutureTests extends ESTestCase {
         assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
         final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
         final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
-        final var threadCount = 4;
         final var running = new AtomicBoolean(true);
-        final var startBarrier = new CyclicBarrier(threadCount + 1);
         try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
-            assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
+            // We only need 4 threads to reproduce this issue reliably, using more threads
+            // just increases the run time due to the additional synchronisation
+            final var threadCount = Math.min(threadPool.info(executorName).getMax(), 4);
+            final var startBarrier = new CyclicBarrier(threadCount + 1);
             // N threads competing to complete the futures
             for (int i = 0; i < threadCount; i++) {
-                threadPool.executor(executorName).submit(() -> {
+                threadPool.executor(executorName).execute(() -> {
                     safeAwait(startBarrier);
                     while (running.get()) {
                         futureReference.get().onResponse(null);
                     }
-                }, null);
+                });
             }
-            // create new futures to complete
+            // The race can only occur once per completion, so we provide
+            // a stream of new futures to the competing threads to
+            // maximise the probability it occurs. Providing them
+            // with new futures while they spin proved to be much
+            // more reliable at reproducing the issue than releasing
+            // them all from a barrier to complete a single future.
             safeAwait(startBarrier);
             for (int i = 0; i < 20; i++) {
                 futureReference.set(new PlainActionFuture<>());

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -146,7 +146,6 @@ public class PlainActionFutureTests extends ESTestCase {
     }
 
     public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
-        assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
         final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
         final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
         final var running = new AtomicBoolean(true);

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -18,11 +18,9 @@ import org.elasticsearch.threadpool.TestThreadPool;
 import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;
 
-import java.util.ArrayList;
 import java.util.concurrent.CancellationException;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
 import java.util.concurrent.atomic.AtomicReference;
@@ -154,27 +152,24 @@ public class PlainActionFutureTests extends ESTestCase {
         final var threadCount = 4;
         final var running = new AtomicBoolean(true);
         final var startBarrier = new CyclicBarrier(threadCount + 1);
-        final var futures = new ArrayList<Future<?>>();
         try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
             assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
-            // N threads competing to complete the future
+            // N threads competing to complete the futures
             for (int i = 0; i < threadCount; i++) {
-                futures.add(threadPool.executor(executorName).submit(() -> {
+                threadPool.executor(executorName).submit(() -> {
                     safeAwait(startBarrier);
                     while (running.get()) {
                         futureReference.get().onResponse(null);
                     }
-                }, null));
+                }, null);
             }
-            // continually create new futures to complete
+            // create new futures to complete
             safeAwait(startBarrier);
-            long endTime = System.currentTimeMillis() + 20;
-            while (System.currentTimeMillis() < endTime) {
+            for (int i = 0; i < 20; i++) {
                 futureReference.set(new PlainActionFuture<>());
                 safeSleep(1);
             }
             running.set(false);
-            futures.forEach(ESTestCase::safeGet);
         }
     }
 

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -19,12 +19,9 @@ import org.elasticsearch.threadpool.ThreadPool;
 import org.elasticsearch.transport.RemoteTransportException;
 
 import java.util.ArrayList;
-import java.util.List;
 import java.util.concurrent.CancellationException;
-import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.CyclicBarrier;
 import java.util.concurrent.ExecutionException;
-import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Future;
 import java.util.concurrent.TimeUnit;
 import java.util.concurrent.atomic.AtomicBoolean;
@@ -152,33 +149,30 @@ public class PlainActionFutureTests extends ESTestCase {
 
     public void testAssertCompleteAllowedAllowsConcurrentCompletesFromSamePool() {
         assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
+        final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
+        final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
+        final var threadCount = 4;
+        final var running = new AtomicBoolean(true);
+        final var startBarrier = new CyclicBarrier(threadCount + 1);
+        final var futures = new ArrayList<Future<?>>();
         try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
-            final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
-            final var executorName = randomFrom(ThreadPool.Names.GENERIC, ThreadPool.Names.MANAGEMENT);
-            final var threadCount = threadPool.info(executorName).getMax();
-            final AtomicBoolean running = new AtomicBoolean(true);
-            final CountDownLatch startLatch = new CountDownLatch(threadCount + 1);
-            final List<Future<?>> futures = new ArrayList<>();
-            final ExecutorService executor = threadPool.executor(executorName);
-            for (int i = 1; i < threadCount; i++) {
-                futures.add(executor.submit(() -> {
-                    startLatch.countDown();
-                    safeAwait(startLatch);
+            assert threadPool.info(executorName).getMax() >= threadCount : "Not enough threads in pool";
+            // N threads competing to complete the future
+            for (int i = 0; i < threadCount; i++) {
+                futures.add(threadPool.executor(executorName).submit(() -> {
+                    safeAwait(startBarrier);
                     while (running.get()) {
                         futureReference.get().onResponse(null);
                     }
                 }, null));
             }
-            executor.submit(() -> {
-                startLatch.countDown();
-                safeAwait(startLatch);
-                while (running.get()) {
-                    futureReference.set(new PlainActionFuture<>());
-                }
-            }, null);
-            startLatch.countDown();
-            safeAwait(startLatch);
-            safeSleep(20);
+            // continually create new futures to complete
+            safeAwait(startBarrier);
+            long endTime = System.currentTimeMillis() + 20;
+            while (System.currentTimeMillis() < endTime) {
+                futureReference.set(new PlainActionFuture<>());
+                safeSleep(1);
+            }
             running.set(false);
             futures.forEach(ESTestCase::safeGet);
         }

--- a/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
+++ b/server/src/test/java/org/elasticsearch/action/support/PlainActionFutureTests.java
@@ -152,19 +152,19 @@ public class PlainActionFutureTests extends ESTestCase {
 
     public void testAssertCompleteAllowedAllowsConcurrentCompletes() throws InterruptedException {
         assumeTrue("Assertions need to be enabled for assertCompleteAllowed to be invoked", Assertions.ENABLED);
-        try (TestThreadPool tp = new TestThreadPool(getTestName())) {
+        try (TestThreadPool threadPool = new TestThreadPool(getTestName())) {
             final AtomicReference<PlainActionFuture<?>> futureReference = new AtomicReference<>(new PlainActionFuture<>());
             final int threads = 4;
             final AtomicBoolean running = new AtomicBoolean(true);
-            final CountDownLatch startBarrier = new CountDownLatch(threads + 1);
-            final CompletionService<?> cs = new ExecutorCompletionService<>(tp.generic());
+            final CountDownLatch startLatch = new CountDownLatch(threads + 1);
+            final CompletionService<?> completionService = new ExecutorCompletionService<>(threadPool.generic());
             final List<Future<?>> futures = new ArrayList<>();
             for (int i = 0; i < threads; i++) {
-                futures.add(cs.submit(new FutureCompleter(i, startBarrier, futureReference, running), null));
+                futures.add(completionService.submit(new FutureCompleter(i, startLatch, futureReference, running), null));
             }
-            startBarrier.countDown();
-            safeAwait(startBarrier);
-            cs.poll(20, TimeUnit.MILLISECONDS);
+            startLatch.countDown();
+            safeAwait(startLatch);
+            completionService.poll(20, TimeUnit.MILLISECONDS);
             running.set(false);
             futures.forEach(ESTestCase::safeGet);
         }

